components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/username/NotificationUsernameRecoveryManager.java

```diff
@@ -209,3 +101,138 @@
+                        if (users.size() > 0) {
+                            resultedUserList = new ArrayList<>(users.size());
+                            resultedUserList.addAll(users);
+                            if (log.isDebugEnabled()) {
+                                log.debug("Current matching temporary userlist :"
+                                        + resultedUserList.toString());
+                            }
+                        } else {
+                            if (log.isDebugEnabled()) {
+                                log.debug("There are no users for " + claim.getClaimURI() + " with the value : "
+                                        + claim.getClaimValue() + " in the previously filtered user list");
+                            }
+                            return false;
+                        }
+                    } else {
+                        resultedUserList.addAll(matchedUserList);
+                        if (log.isDebugEnabled()) {
+                            log.debug("Current matching temporary userlist :"
+                                    + resultedUserList.toString());
+                        }
+                    }
+
+                } else {
+                    if (log.isDebugEnabled()) {
+                        log.debug("There are no matching users for " + claim.getClaimURI() + " with the value : "
+                                + claim.getClaimValue());
+                    }
+                    return false;
+//                    throw Utils.handleClientException(
+//                            IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_NO_USER_FOUND_FOR_RECOVERY, null);
+                }
+            }
+        }
+
+        if (resultedUserList.size() == 1) {
+            user = resultedUserList.get(0);
+            if (log.isDebugEnabled()) {
+                log.debug("There are more than one user in the result set : "
+                        + user.toString());
+            }
+
+            //TODO send the username to the user as an email
+            return true;
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("There are more than one user in the result set : "
+                        + resultedUserList.toString());
+            }
+            return false;
+
+        }
+
+
+    }
+
+    private static List<User> getUserList(String claimUri, String value) throws IdentityRecoveryException {
+
+        RealmService realmService = IdentityRecoveryServiceDataHolder.getInstance().getRealmService();
+        IdentityStore identityStore = realmService.getIdentityStore();
+
+        Claim claim = new Claim();
+        claim.setClaimUri(claimUri);
+        claim.setValue(value);
+        claim.setDialectUri("http://wso2.org/claims");
+
+        try {
+            return identityStore.listUsers(claim, 0, 100);
+        } catch (IdentityStoreException e) {
+            String msg = "Unable to retrieve the user list from claim";
```

Comment:
Would be better if we could add the claim uri and/or claim value to the exception message